pkg/scte35/stream.go

```diff
@@ -0,0 +1,354 @@
+    bufSz = 13000 * pktSz
+)
+
+type PacketData struct {
+	PacketNumber int     `json:",omitempty"`
+	Pid          uint16  `json:",omitempty"`
+	Program      uint16  `json:",omitempty"`
+	Pcr          float64 `json:",omitempty"`
+	Pts          float64 `json:",omitempty"`
+}
+
+//Stream for parsing MPEGTS for SCTE-35
+type Stream struct {
+	pktNum   int // packet count.
+	programs []uint16
+	pid2Prgm map[uint16]uint16 //lookup table for pid to program
+	prgm2pcr map[uint16]uint64 //lookup table for program to pcr
+	prgm2pts map[uint16]uint64 //lookup table for program to pts
+	partial  map[uint16][]byte // partial manages tables spread across multiple packets by pid
+	last     map[uint16][]byte // last compares current packet payload to last packet payload by pid
+	Pids
+}
+
+func (stream *Stream) mkMaps() {
+	stream.pid2Prgm = make(map[uint16]uint16)
+	stream.last = make(map[uint16][]byte)
+	stream.partial = make(map[uint16][]byte)
+	stream.prgm2pcr = make(map[uint16]uint64)
+	stream.prgm2pts = make(map[uint16]uint64)
+}
+
+// Decode fname (a file name) for SCTE-35
+func (stream *Stream) Decode(fname string) {
+	stream.mkMaps()
+	stream.pktNum = 0
+	file, err := os.Open(fname)
+	chk(err)
+	defer file.Close()
+	buffer := make([]byte, bufSz)
+	for {
+		bytesread, err := file.Read(buffer)
+		if err != nil {
+			break
+		}
+		for i := 1; i <= (bytesread / pktSz); i++ {
+			end := i * pktSz
+			start := end - pktSz
+			p := buffer[start:end]
+			pkt := &p
+			stream.pktNum++
+			stream.parse(*pkt)
+		}
+	}
+}
+
+func (stream *Stream) mkPcr(prgm uint16) float64 {
+	pcrb := stream.prgm2pcr[prgm]
+	return mk90k(pcrb)
+}
+
+func (stream *Stream) mkPts(prgm uint16) float64 {
+	pts := stream.prgm2pts[prgm]
+	return mk90k(pts)
+}
+
+func (stream *Stream) parsePusi(pkt []byte) bool {
+	if (pkt[1]>>6)&1 == 1 {
+		if pkt[6]&1 == 1 {
+			return true
+		}
+	}
+	return false
+}
+
+func (stream *Stream) parsePts(pkt []byte, pid uint16) {
+	if stream.parsePusi(pkt) {
+		prgm, ok := stream.pid2Prgm[pid]
+		if ok {
+			pts := (uint64(pkt[13]) >> 1 & 7) << 30
+			pts |= uint64(pkt[14]) << 22
+			pts |= (uint64(pkt[15]) >> 1) << 15
+			pts |= uint64(pkt[16]) << 7
+			pts |= uint64(pkt[17]) >> 1
+			stream.prgm2pts[prgm] = pts
+		}
+	}
+}
+
+//
+func (stream *Stream) parsePcr(pkt []byte, pid uint16) {
+	if (pkt[3]>>5)&1 == 1 {
+		if (pkt[5]>>4)&1 == 1 {
+			pcr := (uint64(pkt[6]) << 25)
+			pcr |= (uint64(pkt[7]) << 17)
+			pcr |= (uint64(pkt[8]) << 9)
+			pcr |= (uint64(pkt[9]) << 1)
+			pcr |= uint64(pkt[10]) >> 7
+			prgm := stream.pid2Prgm[pid]
+			stream.prgm2pcr[prgm] = pcr
+		}
+	}
+}
+
+//parsePay packet payload starts after header and afc (if present)
+func (stream *Stream) parsePayload(pkt []byte) []byte {
+	head := 4
+	hasafc := (pkt[3] >> 5) & 1
+	if hasafc == 1 {
+		afl := int(pkt[4])
+		head += afl + 1
+	}
+	if head > pktSz {
+		head = pktSz
+	}
+	return pkt[head:]
+}
+
+//chkPartial appends the current packet payload to partial table by pid.
```

Comment:
When adding comments, there should be a space between `//` and the comment itself (bit pedantic but my linter shouts at me when we violate this)